src/ucp/core/ucp_ep.c

```diff
@@ -3496,14 +3496,21 @@ void ucp_ep_req_purge(ucp_ep_h ucp_ep, ucp_request_t *req,
         ucp_request_complete_send(req, status);
         ucp_ep_rma_remote_request_completed(ucp_ep);
     } else {
-        /* SW RMA/PUT and AMO/Post operations don't allocate local request ID
-         * and don't need to be tracked, since they complete UCP request upon
-         * sending all data to a peer. Receiving RMA/CMPL and AMO/REP packets
-         * complete flush requests */
-        ucs_assert((req->send.uct.func != ucp_rma_sw_proto.progress_put) &&
-                   (req->send.uct.func != ucp_amo_sw_proto.progress_post));
         ucs_assert(req->send.ep == ucp_ep);
 
+        if (req->send.uct.func == ucp_proto_progress_rndv_rtr) {
+            if (req->send.rndv.mdesc != NULL) {
```

Comment:
it seems that ucp_rndv_receive->ucp_rndv_req_send_rtr does not initialize desc to NULL

Reply:
yeah, it wrongly set by a caller
[`        frndv_req->send.rndv.mdesc   = mdesc;`](https://github.com/openucx/ucx/blob/7ea28295471b18f691ec449d39dee97bd11e6553/src/ucp/rndv/rndv.c#L1405)
which uses other union part of send request `send.rndv` but `ucp_proto_progress_rndv_rtr` uses `send.rndv_rtr`

Reply:
maybe we should check RNDV_FRAG request flag?

Reply:
RTR does not have this flag